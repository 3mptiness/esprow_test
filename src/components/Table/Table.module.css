.Container {
  width: 100%;
  height: 100%;
  overflow: auto;
  scroll-behavior: smooth;
}

.Column {
  cursor: pointer;
}

.Sort {
  position: absolute;
}

.Input {
  width: 100%;
  height: 100%;
  box-sizing: border-box;
}
