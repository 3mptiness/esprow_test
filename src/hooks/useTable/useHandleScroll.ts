import { UIEventHandler, useCallback, useState } from 'react';

export const useHandleScroll = (rowHeight: number) => {
	const [scroll, setScroll] = useState(0);

	const handleScroll: UIEventHandler<HTMLDivElement> = useCallback((e) => {
		setScroll((e.target as HTMLDivElement).scrollTop);
	}, []);

	const startRowIndex = Math.trunc((scroll - 25) / rowHeight);

	return {
		scroll: startRowIndex * rowHeight,
		startRowIndex,
		handleScroll,
	};
};
