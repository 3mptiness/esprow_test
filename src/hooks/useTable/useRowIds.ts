import { useMemo } from 'react';
import { generateUuid } from '../../utils/generateUuid';
import { RowType } from '../../types';

export const useRowIds = (defaultRows: RowType[]) =>
	useMemo(
		() =>
			new WeakMap(
				defaultRows.map(row => [
					row,
					row.id && typeof row.id === 'string' ? row.id : generateUuid(),
				])
			),
		[defaultRows]
	);
