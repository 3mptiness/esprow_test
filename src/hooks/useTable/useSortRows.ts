import { useCallback, useMemo, useState } from 'react';
import { HeadersMap, RowType } from '../../types';

type UseSortRowsProps = { defaultRows: RowType[]; headers: HeadersMap };

export const useSortRows = (props: UseSortRowsProps) => {
	const { headers, defaultRows } = props;

	const [[sortKey, sortDesc], setSort] = useState<
		[] | [string] | [string, boolean]
	>([]);

	const sortedRows = useMemo(() => {
		if (!sortKey) return defaultRows;

		const type = headers.get(sortKey);
		const isString = ['input', 'textarea', 'email', 'date', 'radio'].includes(
			type ?? ''
		);

		if (isString) {
			return [...defaultRows].sort((rowA: RowType, rowB: RowType) => {
				const a = rowA[sortKey] as string;
				const b = rowB[sortKey] as string;
				return sortDesc ? b.localeCompare(a) : a.localeCompare(b);
			});
		}

		return [...defaultRows].sort((rowA: RowType, rowB: RowType) => {
			const a = rowA[sortKey] as number;
			const b = rowB[sortKey] as number;
			return sortDesc ? b - a : a - b;
		});
	}, [headers, defaultRows, sortKey, sortDesc]);

	const toggleSort = useCallback(
		(key: string) => {
			if (key !== sortKey) return setSort([key]);

			if (sortDesc) {
				return setSort([]);
			}

			return setSort([key, true]);
		},
		[sortDesc, sortKey]
	);

	const getSortArrow = useCallback(
		(title: string) => {
			if (!sortKey || sortKey !== title) return '';

			return sortDesc ? '▲' : '▼';
		},
		[sortKey, sortDesc]
	);

	return {
		getSortArrow,
		toggleSort,
		sortedRows,
	};
};
