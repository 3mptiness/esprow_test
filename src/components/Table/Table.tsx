import React, { FC } from 'react';
import { useTable } from '../../hooks/useTable';
import { TableProps } from '../../types';
import css from './Table.module.css';
import { THead } from './THead/THead';
import { TBody } from './TBody/TBody';
import { Row } from './Row';

export const Table: FC<TableProps> = (props) => {
	const {
		headers,
		rows,
		handleChange,
		handleScroll,
		ref,
		toggleSort,
		getSortArrow,
		fullHeight,
		rowHeight,
		scroll,
		rowIds,
	} = useTable(props);

	return (
		<div className={css.Container} ref={ref} onScroll={handleScroll}>
			<table>
				<THead
					headers={headers}
					getSortArrow={getSortArrow}
					toggleSort={toggleSort}
				/>
				<TBody
					fullHeight={fullHeight}
					rowHeight={rowHeight}
					scroll={scroll}
					rowsLength={rows.length}
				>
					{rows.map((row, index) => {
						const rowId = rowIds.get(row) as string;
						return (
							<Row
								key={rowId}
								id={rowId}
								index={index}
								row={row}
								headers={headers}
								onChange={handleChange}
								height={rowHeight}
							/>
						);
					})}
				</TBody>
			</table>
		</div>
	);
};
