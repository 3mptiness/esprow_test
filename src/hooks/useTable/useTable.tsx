import { useCallback, useMemo } from 'react';
import { getHeadersFromRows } from '../../utils/getHeadersFromRows';
import { ChangeHandler, RowType, TableProps } from '../../types';
import { useSize } from '../useSize';
import { DEFAULT_ROW_HEIGHT } from '../../constants';
import { getRowsFromJsonArray } from '../../utils/getRowsFromJsonArray';
import { useHandleScroll } from './useHandleScroll';
import { useSortRows } from './useSortRows';
import { useRowIds } from './useRowIds';

export const useTable = (props: TableProps) => {
	const { jsonArray, onChange, rowHeight = DEFAULT_ROW_HEIGHT } = props;

	const { ref, height } = useSize();

	const defaultRows: RowType[] = useMemo(
		() => getRowsFromJsonArray(jsonArray),
		// 	no need recalculate every update of jsonArray
		// eslint-disable-next-line react-hooks/exhaustive-deps
		[jsonArray.length]
	);

	const headers = useMemo(() => getHeadersFromRows(defaultRows), [defaultRows]);

	const { sortedRows, toggleSort, getSortArrow } = useSortRows({
		defaultRows,
		headers,
	});

	const { startRowIndex, scroll, handleScroll } = useHandleScroll(rowHeight);

	const rows = useMemo(() => {
		const rowsInViewport = Math.ceil(height / rowHeight);
		return sortedRows.slice(startRowIndex, startRowIndex + rowsInViewport);
	}, [height, rowHeight, sortedRows, startRowIndex]);

	const handleChange: ChangeHandler = useCallback(
		({ index, prop, value }) => {
			const changedRow = sortedRows[index];
			const temp = [...defaultRows];
			const indexOfChanged = temp.indexOf(changedRow);
			temp[indexOfChanged] = { ...changedRow, [prop]: value };
			onChange && onChange(temp);
		},
		[sortedRows, defaultRows, onChange]
	);

	return {
		ref,
		rowHeight,
		headers,
		scroll,
		handleScroll,
		rows,
		toggleSort,
		getSortArrow,
		rowIds: useRowIds(defaultRows),
		fullHeight: rowHeight * defaultRows.length,
		handleChange,
	};
};
