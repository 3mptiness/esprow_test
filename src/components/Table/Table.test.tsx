import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { Table } from './Table';
import { useTable } from '../../hooks/useTable';

jest.mock('../../hooks/useTable');

describe('Table', () => {
	it('success render', () => {
		const row = { testKey: 'test data', otherKey: 'other data' };
		(useTable as jest.Mock).mockReturnValue({
			rows: [row],
			rowIds: new WeakMap([[row, 'id']]),
			headers: new Map([['testKey', 'input']]),
			handleChange: jest.fn(),
			handleScroll: jest.fn(),
			toggleSort: jest.fn(),
			getSortArrow: jest.fn(),
			fullHeight: 80,
			rowHeight: 80,
			scroll: 0,
		});
		expect(
			createRenderer().render(<Table jsonArray={[row]} />)
		).toMatchSnapshot();
	});
});
