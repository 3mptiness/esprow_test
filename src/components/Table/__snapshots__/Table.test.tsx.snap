// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table success render 1`] = `
<div
  className="Container"
  onScroll={[MockFunction]}
>
  <table>
    <Memo
      getSortArrow={[MockFunction]}
      headers={
        Map {
          "testKey" => "input",
        }
      }
      toggleSort={[MockFunction]}
    />
    <Memo
      fullHeight={80}
      rowHeight={80}
      rowsLength={1}
      scroll={0}
    >
      <Memo
        headers={
          Map {
            "testKey" => "input",
          }
        }
        height={80}
        id="id"
        index={0}
        onChange={[MockFunction]}
        row={
          Object {
            "otherKey": "other data",
            "testKey": "test data",
          }
        }
      />
    </Memo>
  </table>
</div>
`;
